### ISSUE TITLE
Можно ли вынести bemtree шаблоны на клиент?

### ISSUE THREAD

**Issue by 4ok** [bem/bem-forum-content-ru#223]
Сейчас в основном, все что не касается html (тэги, атрибуты) я обрабатываю в bemhtml, остальное в bemtree. А если мне понадобятся эти же преобразования на клиенте, то их надо будет перенести в bemhtml или могу шаблоны bemtree по депсам включить в собираемый js? Или все же bemtree это чисто серверная технология?

К примеру одно из обычных преобразований:

 
```js
block('custom-menu').elem('item')
(
	def()(function () {
		var ctx = this.ctx;

		return applyCtx({
			block : 'link',
			url   : ctx.url
			mix   : {
				block: this.block,
				elem:  this.elem
			}
			content : ctx.content
		})
	})
)
```

Вот, если это преобразование необходмо на клиенте оно должно быть в bemhtml? Просто по логике мне хочется пред. код описать в bemtree, а всякие теги и атрибуты для этого блока в bemhtml.

**Comment by apsavin** [comment]
bemtree - это просто шаблоны, нет ничего плохого в шаблонах на клиенте.

**Comment by zxqfox** [comment]
Более того — для Single Page Application (или близкой к SPA архитектуре) в них есть еще и куча плюсов ;-)

**Comment by 4ok** [comment]
Так их можно подключить на клиент в общий js? Если так, то теряется смысл моды def() и content() для bemhtml, т.к преобразование по этим модам необходимо выполнять в bemtree.

**Comment by tadatuta** [comment]
@4ok безусловно их можно подключать в общий js.
`def()` в BEMTHML действительно чаще всего не требуется, а `content()` нужен для кейсов создания число оформительских оберток и добавления всяких визуальных штук в разметку, которые не несут особой семантической нагрузки.

**Comment by 4ok** [comment]
Приведи пример пож-то визуальных штук, чтобы было ясно, что именно имеется ввиду, спасибо.

**Comment by tadatuta** [comment]
@4ok например, элемент для иконки, какие-то разделители и прочие украшательства, которые по каким-то причинам не удается выразить через `:before` и `:after` в CSS

**Comment by 4ok** [comment]
@tadatuta, спасибо, понял.

**Comment by 4ok** [comment]
В bemtree шаблонах на сервере я часто использую require необходимых модулей, выходит, что в случае подключения этих шаблонов на клиент, необходимо придумать как обходится без них. Что можно придумать в этом случае?

Например необходим модуль форматирования даты - для ноды это require в самом bemtree, а на клиенте подключение через `<script src="..."></script>`. Но в этом случае сам require в шаблоне я так понимаю нужно удалить и передать этот модуль как параметр в bemtree.

Правильное направление мысли?

**Comment by tadatuta** [comment]
@4ok можно заворачивать все в YModules, которая работает и на клиенте и на сервере.

Получится что-то типа:

```js
// m1.node.js
modules.define('m1', function(provide) {
    provide(require('path/to/something');
});
```

```js
// m1.browser.js
modules.define('m1', function(provide) {
    provide(/*borschik:include:path/to/something*/);
});
```

**Comment by 4ok** [comment]
@tadatuta - супер! Но как потом эти объявленные модули использовать в bemtree? Не соображу никак.

**Comment by tadatuta** [comment]
@4ok например, вот так:

```js
modules.require(['m1', 'BEMTREE'], function(M1, BEMTREE) {
    BEMTREE.apply({ block: 'page', modules: { m1: M1 } });
});
```

дальше можно пробросить все необходимые модули вглубь по дереву в шаблонах:
```js
block('page').def()(function() {
    return applyNext({ modules: this.ctx.modules });
});
```

и, соответственно, они будут доступны из всех блоков:

```js
block('my-block').content()(function() {
    return this.modules.m1('trololo');
});
```

**Comment by Guria** [comment]
Я вот подобное в `BH` изобразил для инлайновых svg.
```js
if (require) {
  require('./kg-glyph.prepare.js'); // prepare json from folder with svg
  var glyphs = require('./glyphs.json'); 
} else {
  var glyphs = {/*borschik:include:../../common.blocks/kg-glyph/glyphs.json*/};
}
module.exports = function (bh) {
  bh.match('kg-glyph', function(ctx, json) {
    return glyphs[json.glyph+'.svg'];
  });
};
```
@tadatuta а на сколько адекватно использовать пободную конструкцию?

**Comment by tadatuta** [comment]
@Guria Нормально )

Я бы еще предложил посмотреть в сторону инлайнинга SVG прямо средствами `borschik`:

```js
{
    "freeze_paths" : {
        "libs/**": ":base64:",
        "libs/**/*.svg": ":encodeURIComponent:",
        "*.blocks/**": ":base64:",
        "*.blocks/**/*.svg": ":encodeURIComponent:"
    }
}
```

+ https://ru.bem.info/articles/borschik/#Динамический-url

**Comment by 4ok** [comment]
Хотелось бы тогда технологии делить на серверные, клиентские и общие по суффиксу. Правильно ли разделить их след. образом -  *.node.bemtree, *.browser.bemtree, *.bemtree?

**Comment by zxqfox** [comment]
В силу того, что мы можем доопределять функциональность на разных уровнях и нас никто не может оставить от этого — не проще вынести общий

**Comment by tadatuta** [comment]
@4ok да, вполне можно делить так (для консистентности можно еще последовать принятой в bem-core схеме и общее для всех платформ явно назвать `vanilla.bemtree`), а можно и разнести по уровням, как предлагает @zxqfox.
Но я бы рекомендовал не плодить сущностей и ограничиться просто использованием `depsByTech`, если только не предполагается, что будет много блоков, которые принципиально несовместимы, но называются одинаково.

**Comment by 4ok** [comment]
@tadatuta, @zxqfox - доступно, спасибо.
